In [28]:
from googleapi import *
import pprint
import pandas as pd
import os
from config import DATA_DIR
pd.set_option("display.max.columns", None)

In [ ]:
COPENHAGEN_BOUNDS = {
    "lat_south": 55.51,
    "lat_north": 55.82,
    "lon_west": 12.23,
    "lon_east": 12.73,
}

In [12]:
responseUsingRegionFilter = find_aggregated_places("", useRegionFilter=True)
pprint.pp(responseUsingRegionFilter)

Elapsed time: 953.38004 milliseconds for thread MAKE_REQUEST(URL='HTTPS://AREAINSIGHTS.GOOGLEAPIS.COM/V1:COMPUTEINSIGHTS', BODY={'INSIGHTS': ['INSIGHT_COUNT', 'INSIGHT_PLACES'], 'FILTER': {'LOCATIONFILTER': {'REGION': {'PLACE': 'PLACES/CHIJIZ2AXDXTUKYRUGEU5T1-3QQ'}}, 'TYPEFILTER': {'INCLUDEDTYPES': 'RESTAURANT'}}})
Elapsed time: 953.60279 milliseconds for thread FIND_AGGREGATED_PLACES(COORDINATES='', USEREGIONFILTER=TRUE)
{'error': {'code': 429,
           'message': 'The response exceeded the maximum number of allowed '
                      'places. Only 100 places can be returned in a single '
                      'response. The request produced 2652 places and '
                      'therefore no places were returned.',
           'status': 'RESOURCE_EXHAUSTED'}}


# Bounding box

Coordinates for copenhagen in a bounding box:

``` bash
(topleft):
    55.82 12.23 
(bottomright):
    55.51 12.73
```

The strange thing is, that the response of the call with the same coordinates, isn't consistent. See below

In [11]:
responseUsingCoordinateBounds = find_aggregated_places(COPENHAGEN_BOUNDS) 
pprint.pp(responseUsingCoordinateBounds)

Elapsed time: 1.03069 seconds for thread MAKE_REQUEST(URL='HTTPS://AREAINSIGHTS.GOOGLEAPIS.COM/V1:COMPUTEINSIGHTS', BODY={'INSIGHTS': ['INSIGHT_COUNT', 'INSIGHT_PLACES'], 'FILTER': {'LOCATIONFILTER': {'CUSTOMAREA': {'POLYGON': {'COORDINATES': [{'LATITUDE': 55.82, 'LONGITUDE': 12.23}, {'LATITUDE': 55.51, 'LONGITUDE': 12.23}, {'LATITUDE': 55.51, 'LONGITUDE': 12.73}, {'LATITUDE': 55.82, 'LONGITUDE': 12.73}, {'LATITUDE': 55.82, 'LONGITUDE': 12.23}]}}}, 'TYPEFILTER': {'INCLUDEDTYPES': 'RESTAURANT'}}})
Elapsed time: 1.03099 seconds for thread FIND_AGGREGATED_PLACES(COORDINATES={'LAT_SOUTH': 55.51, 'LAT_NORTH': 55.82, 'LON_WEST': 12.23, 'LON_EAST': 12.73})
{'error': {'code': 429,
           'message': 'The response exceeded the maximum number of allowed '
                      'places. Only 100 places can be returned in a single '
                      'response. The request produced 3971 places and '
                      'therefore no places were returned.',
           'status': 'RESOURCE_E

In [14]:
responseUsingCoordinateBounds = find_aggregated_places(COPENHAGEN_BOUNDS) 
pprint.pp(responseUsingCoordinateBounds)

Elapsed time: 1.00002 seconds for thread MAKE_REQUEST(URL='HTTPS://AREAINSIGHTS.GOOGLEAPIS.COM/V1:COMPUTEINSIGHTS', BODY={'INSIGHTS': ['INSIGHT_COUNT', 'INSIGHT_PLACES'], 'FILTER': {'LOCATIONFILTER': {'CUSTOMAREA': {'POLYGON': {'COORDINATES': [{'LATITUDE': 55.82, 'LONGITUDE': 12.23}, {'LATITUDE': 55.51, 'LONGITUDE': 12.23}, {'LATITUDE': 55.51, 'LONGITUDE': 12.73}, {'LATITUDE': 55.82, 'LONGITUDE': 12.73}, {'LATITUDE': 55.82, 'LONGITUDE': 12.23}]}}}, 'TYPEFILTER': {'INCLUDEDTYPES': 'RESTAURANT'}}})
Elapsed time: 1.00048 seconds for thread FIND_AGGREGATED_PLACES(COORDINATES={'LAT_SOUTH': 55.51, 'LAT_NORTH': 55.82, 'LON_WEST': 12.23, 'LON_EAST': 12.73})
{'error': {'code': 429,
           'message': 'The response exceeded the maximum number of allowed '
                      'places. Only 100 places can be returned in a single '
                      'response. The request produced 4221 places and '
                      'therefore no places were returned.',
           'status': 'RESOURCE_E

In [24]:
df_lim20 = pd.read_json(DATA_DIR / "copenhagen-bounds-limit20_20260312_115801.json")
df_lim10 = pd.read_json(DATA_DIR / "copenhagen-bounds-limit10_20260316_125502.json")

In [ ]:
df_lim10.shape

(4392, 66)

In [26]:
df_lim20.shape

(4364, 66)

In [27]:
df_lim10.info()

<class 'pandas.DataFrame'>
RangeIndex: 4392 entries, 0 to 4391
Data columns (total 66 columns):
 #   Column                        Non-Null Count  Dtype  
---  ------                        --------------  -----  
 0   name                          4392 non-null   str    
 1   id                            4392 non-null   str    
 2   types                         4392 non-null   object 
 3   nationalPhoneNumber           3758 non-null   str    
 4   internationalPhoneNumber      3758 non-null   str    
 5   formattedAddress              4392 non-null   str    
 6   addressComponents             4392 non-null   object 
 7   plusCode                      4381 non-null   object 
 8   location                      4392 non-null   object 
 9   viewport                      4392 non-null   object 
 10  rating                        4046 non-null   float64
 11  googleMapsUri                 4392 non-null   str    
 12  websiteUri                    3600 non-null   str    
 13  regularOpening

In [29]:
df_lim10.head()

,name,id,types,nationalPhoneNumber,internationalPhoneNumber,formattedAddress,addressComponents,plusCode,location,viewport,rating,googleMapsUri,websiteUri,regularOpeningHours,utcOffsetMinutes,adrFormatAddress,businessStatus,priceLevel,userRatingCount,iconMaskBaseUri,iconBackgroundColor,displayName,primaryTypeDisplayName,takeout,delivery,dineIn,reservable,servesBreakfast,servesLunch,servesDinner,servesBeer,servesWine,servesBrunch,currentOpeningHours,currentSecondaryOpeningHours,regularSecondaryOpeningHours,primaryType,shortFormattedAddress,reviews,photos,outdoorSeating,liveMusic,menuForChildren,servesCocktails,servesDessert,servesCoffee,goodForChildren,allowsDogs,restroom,goodForGroups,goodForWatchingSports,paymentOptions,parkingOptions,accessibilityOptions,addressDescriptor,googleMapsLinks,priceRange,timeZone,postalAddress,googleMapsTypeLabel,curbsidePickup,containingPlaces,editorialSummary,fuelOptions,servesVegetarianFood,consumerAlert
0,places/ChIJ6eXKTWdDUkYR7k5XjV6Ju-0,ChIJ6eXKTWdDUkYR7k5XjV6Ju-0,"[restaurant, inn, food, lodging, point_of_inte...",48 18 30 01,+45 48 18 30 01,"Bygaden 30, 3660 Stenløse, Denmark","[{'longText': '30', 'shortText': '30', 'types'...","{'globalCode': '9F7JQ7V7+94', 'compoundCode': ...","{'latitude': 55.793493999999995, 'longitude': ...","{'low': {'latitude': 55.792122919708504, 'long...",4.4,https://maps.google.com/?cid=17130436646825053...,http://ganlosekro.dk/,"{'openNow': False, 'periods': [{'open': {'day'...",60,"<span class=""street-address"">Bygaden 30</span>...",OPERATIONAL,PRICE_LEVEL_MODERATE,1376.0,https://maps.gstatic.com/mapfiles/place_api/ic...,#FF9E67,"{'text': 'Ganløse Kro', 'languageCode': 'da'}","{'text': 'Restaurant', 'languageCode': 'en-US'}",1.0,0.0,1.0,1.0,0.0,1.0,1.0,1.0,1.0,1.0,"{'openNow': False, 'periods': [{'open': {'day'...","[{'openNow': False, 'periods': [{'open': {'day...","[{'openNow': False, 'periods': [{'open': {'day...",restaurant,"Bygaden 30, Stenløse",[{'name': 'places/ChIJ6eXKTWdDUkYR7k5XjV6Ju-0/...,[{'name': 'places/ChIJ6eXKTWdDUkYR7k5XjV6Ju-0/...,0.0,1.0,1.0,1.0,1.0,1.0,1.0,0.0,1.0,1.0,0.0,"{'acceptsCreditCards': True, 'acceptsDebitCard...","{'freeParkingLot': True, 'freeStreetParking': ...","{'wheelchairAccessibleParking': True, 'wheelch...",{'landmarks': [{'name': 'places/ChIJdYNxqmBDUk...,{'directionsUri': 'https://www.google.com/maps...,"{'startPrice': {'currencyCode': 'DKK', 'units'...",{'id': 'Europe/Copenhagen'},"{'regionCode': 'DK', 'languageCode': 'en-US', ...","{'text': 'Restaurant', 'languageCode': 'en-US'}",NaN,NaN,NaN,NaN,NaN,NaN
1,places/ChIJ21qBwtRDUkYRGBAxnmFfg10,ChIJ21qBwtRDUkYRGBAxnmFfg10,"[restaurant, food, point_of_interest, establis...",93 92 21 42,+45 93 92 21 42,"Bygaden 7, 3660 Stenløse, Denmark","[{'longText': '7', 'shortText': '7', 'types': ...","{'globalCode': '9F7JQ7R7+C8', 'compoundCode': ...","{'latitude': 55.7911146, 'longitude': 12.263354}","{'low': {'latitude': 55.789767219708494, 'long...",4.7,https://maps.google.com/?cid=67383343403468759...,http://www.hjaltesburesoe.dk/,"{'openNow': True, 'periods': [{'open': {'day':...",60,"<span class=""street-address"">Bygaden 7</span>,...",OPERATIONAL,PRICE_LEVEL_MODERATE,139.0,https://maps.gstatic.com/mapfiles/place_api/ic...,#FF9E67,"{'text': 'Hjaltes Ganløse', 'languageCode': 'da'}","{'text': 'Restaurant', 'languageCode': 'en-US'}",1.0,0.0,1.0,1.0,1.0,1.0,1.0,1.0,1.0,1.0,"{'openNow': True, 'periods': [{'open': {'day':...",NaN,NaN,restaurant,"Bygaden 7, Stenløse",[{'name': 'places/ChIJ21qBwtRDUkYRGBAxnmFfg10/...,[{'name': 'places/ChIJ21qBwtRDUkYRGBAxnmFfg10/...,1.0,0.0,1.0,NaN,1.0,1.0,1.0,NaN,1.0,1.0,0.0,"{'acceptsCreditCards': True, 'acceptsDebitCard...","{'freeParkingLot': True, 'freeStreetParking': ...",NaN,{'landmarks': [{'name': 'places/ChIJm0zBFl5DUk...,{'directionsUri': 'https://www.google.com/maps...,"{'startPrice': {'currencyCode': 'DKK', 'units'...",{'id': 'Europe/Copenhagen'},"{'regionCode': 'DK', 'languageCode': 'en-US', ...","{'text': 'Brasserie', 'languageCode': 'en-US'

In [39]:
df_consumer_alert = df_lim10[df_lim10["consumerAlert"].notna()]
df_consumer_alert.displayName

1062    {'text': 'Ramen Hokkaido Nordhavn', 'languageC...
1567          {'text': 'Mexihagen', 'languageCode': 'en'}
Name: displayName, dtype: object

# Resources: Places

In order to understand the response from the call, we rely on [Google's Documentation for Places](https://developers.google.com/maps/documentation/places/web-service/reference/rest/v1/places)

## Google Places API — DataFrame Columns

> Made with assistance from @Claude

| # | Column | Non-Null Count | Dtype | Relevance | Note |
|---|--------|---------------|-------|-----------|------|
| 0 | name | 4392 | str | High | `places/{placeId}` format (not textual name)|
| 1 | id | 4392 | str | Medium | same as `name`, only without prefix `places/`|
| 2 | types | 4392 | object | High | `restaurant`, `meal_takeaway` etc. |
| 3 | nationalPhoneNumber | 3758 | str | Low | 
| 4 | internationalPhoneNumber | 3758 | str | Low |
| 5 | formattedAddress | 4392 | str | Medium | annotation perhaps |
| 6 | addressComponents | 4392 | object | Low | don't think so, *"The structured components that form the formatted address, if this information is available."* |
| 7 | plusCode | 4381 | object | Low | another unit of location. [Plus Codes](https://maps.google.com/pluscodes/) |
| 8 | location | 4392 | object | High | lat lon |
| 9 | viewport | 4392 | object | Low |
| 10 | rating | 4046 | float64 | Medium |
| 11 | googleMapsUri | 4392 | str | High | where we can get the menu data from |
| 12 | websiteUri | 3600 | str | Medium |
| 13 | regularOpeningHours | 3805 | object | Low |
| 14 | utcOffsetMinutes | 4392 | int64 | Low |
| 15 | adrFormatAddress | 4392 | str | Low |
| 16 | businessStatus | 4392 | str | High | `'OPERATIONAL'` or `'CLOSED_TEMPORARILY'`|
| 17 | priceLevel | 1738 | str | Medium |
| 18 | userRatingCount | 4046 | float64 | Low | number of reviews of the place |
| 19 | iconMaskBaseUri | 4392 | str | Low |
| 20 | iconBackgroundColor | 4392 | str | Low |
| 21 | displayName | 4392 | object | High | actual name, fx. Jaggers |
| 22 | primaryTypeDisplayName | 4387 | object | Low | "Pastry Shop", "Restaurant", see `primaryType` col 36: maybe more relevant |
| 23 | takeout | 3237 | float64 | Medium |
| 24 | delivery | 2968 | float64 | Medium |
| 25 | dineIn | 3910 | float64 | Medium |
| 26 | reservable | 2367 | float64 | Low |
| 27 | servesBreakfast | 1732 | float64 | Medium |
| 28 | servesLunch | 3009 | float64 | Medium |
| 29 | servesDinner | 3249 | float64 | Medium |
| 30 | servesBeer | 2754 | float64 | Low |
| 31 | servesWine | 2451 | float64 | Low |
| 32 | servesBrunch | 1713 | float64 | Medium |
| 33 | currentOpeningHours | 3805 | object | Low |
| 34 | currentSecondaryOpeningHours | 543 | object | Low |
| 35 | regularSecondaryOpeningHours | 543 | object | Low |
| 36 | primaryType | 4387 | str | High | "restaurant", "pastry_shop" |
| 37 | shortFormattedAddress | 4392 | str | Low |
| 38 | reviews | 4046 | object | Medium |
| 39 | photos | 3966 | object | Low |
| 40 | outdoorSeating | 2187 | float64 | Low |
| 41 | liveMusic | 3465 | float64 | Low |
| 42 | menuForChildren | 2044 | float64 | Low |
| 43 | servesCocktails | 2035 | float64 | Low |
| 44 | servesDessert | 2428 | float64 | Medium |
| 45 | servesCoffee | 2573 | float64 | Low |
| 46 | goodForChildren | 2214 | float64 | Low |
| 47 | allowsDogs | 1481 | float64 | Low |
| 48 | restroom | 3003 | float64 | Low |
| 49 | goodForGroups | 1847 | float64 | Low |
| 50 | goodForWatchingSports | 3126 | float64 | Low |
| 51 | paymentOptions | 4206 | object | Low |
| 52 | parkingOptions | 1987 | object | Low |
| 53 | accessibilityOptions | 2851 | object | Low |
| 54 | addressDescriptor | 4392 | object | Low | describes other things in the location area |
| 55 | googleMapsLinks | 4392 | object | Low | probably not. `{"directionsUri", "placeUri", "writeAReviewUri", "reviewsUri", "photosUri"}` |
| 56 | priceRange | 3090 | object | Medium |
| 57 | timeZone | 4392 | object | Low |
| 58 | postalAddress | 4386 | object | Medium |
| 59 | googleMapsTypeLabel | 4392 | object | Low | again the same as col 36, but in plaintext |
| 60 | curbsidePickup | 1205 | float64 | Low |
| 61 | containingPlaces | 745 | object | Low |
| 62 | editorialSummary | 344 | object | Medium |
| 63 | fuelOptions | 43 | object | Low |
| 64 | servesVegetarianFood | 1497 | float64 | High |
| 65 | consumerAlert | 2 | object | Low | *"The consumer alert message for the place when we detect suspicious review activity on a business or a business violates our policies."*" - lol |

In [41]:
cols_to_keep = [
    "name", 
    "id", 
    "types", 
    "formattedAddress", 
    "location", 
    "rating", 
    "googleMapsUri",
    "websiteUri",
    "businessStatus",
    "priceLevel",
    "displayName",
    "takeout",
    "delivery",
    "dineIn",
    "servesBreakfast",
    "servesLunch",
    "servesDinner",
    "servesBrunch",
    "servesDessert",
    "primaryType",
    "reviews",
    "priceRange",
    "postalAddress",
    "editorialSummary",
    "servesVegetarianFood",
]

len(cols_to_keep)

25

In [43]:
df_lim10 = df_lim10[cols_to_keep]
df_lim10.info()

<class 'pandas.DataFrame'>
RangeIndex: 4392 entries, 0 to 4391
Data columns (total 25 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   name                  4392 non-null   str    
 1   id                    4392 non-null   str    
 2   types                 4392 non-null   object 
 3   formattedAddress      4392 non-null   str    
 4   location              4392 non-null   object 
 5   rating                4046 non-null   float64
 6   googleMapsUri         4392 non-null   str    
 7   websiteUri            3600 non-null   str    
 8   businessStatus        4392 non-null   str    
 9   priceLevel            1738 non-null   str    
 10  displayName           4392 non-null   object 
 11  takeout               3237 non-null   float64
 12  delivery              2968 non-null   float64
 13  dineIn                3910 non-null   float64
 14  servesBreakfast       1732 non-null   float64
 15  servesLunch           3009 non-n